### PR TITLE
bpf: Fix missing revDNAT when IPsec is used with L7 ingress policy and KPR

### LINKED ISSUE #32897: Broken connectivity of pod to remote nodeport when IPsec is used with L7 ingress policy and KPR
### Is there an existing issue for this?

- [X] I have searched the existing issues

### 1. Create kind cilium cluster 

```

make kind-down
export IMAGE=kindest/node:v1.29.4@sha256:3abb816a5b1061fb15c6e9e60856ec40d56b7b52bcea5f5f1350bc6e2320b6f8
./contrib/scripts/kind.sh --xdp --secondary-network "" 3 "" "" iptables dual 0.0.0.0 6443
 kubectl patch node kind-worker3 --type=json -p='[{"op":"add","path":"/metadata/labels/cilium.io~1no-schedule","value":"true"}]'

  if [[ "gcm(aes)" == "gcm(aes)" ]]; then
    key="rfc4106(gcm(aes)) $(dd if=/dev/urandom count=20 bs=1 2> /dev/null | xxd -p -c 64) 128"
  elif [[ "gcm(aes)" == "cbc(aes)" ]]; then
    key="hmac(sha256) $(dd if=/dev/urandom count=32 bs=1 2> /dev/null| xxd -p -c 64) cbc(aes) $(dd if=/dev/urandom count=32 bs=1 2> /dev/null| xxd -p -c 64)"
  else
    echo "Invalid key type"; exit 1
  fi
  kubectl create -n kube-system secret generic cilium-ipsec-keys \
    --from-literal=keys="3+ ${key}"

git checkout 8cd748de7f4011d4ab9fc04338fa69c194205e6a # https://github.com/cilium/cilium/commit/8cd748de7f4011d4ab9fc04338fa69c194205e6a
make kind-image
kind load --name kind docker-image localhost:5000/cilium/cilium-dev:local
 kind load --name kind docker-image localhost:5000/cilium/operator-generic:local
./cilium-cli install --wait     --chart-directory=./install/kubernetes/cilium     --helm-set=debug.enabled=true     --helm-set=debug.verbose=envoy     --helm-set=hubble.eventBufferCapacity=65535     --helm-set=bpf.monitorAggregation=none     --helm-set=cluster.name=default     --nodes-without-cilium     --helm-set-string=kubeProxyReplacement=true     --set='' --helm-set-string=routingMode=native --helm-set-string=autoDirectNodeRoutes=true --helm-set-string=ipv4NativeRoutingCIDR=10.244.0.0/16 --helm-set-string=ipv6NativeRoutingCIDR=fd00:10:244::/56   --helm-set-string=endpointRoutes.enabled=true --helm-set=ipv6.enabled=true   --helm-set=encryption.enabled=true --helm-set=encryption.type=ipsec --helm-set=encryption.nodeEncryption=false --helm-set=image.repository=localhost:5000/cilium/cilium-dev   --helm-set=image.useDigest=false   --helm-set=image.tag=local   --helm-set=operator.image.repository=localhost:5000/cilium/operator      --helm-set=operator.image.tag=local --helm-set operator.image.suffix=   --helm-set=operator.image.useDigest=false  --helm-set=image.pullPolicy=IfNotPresent --helm-set=operator.image.pullPolicy=IfNotPresent  --helm-set=debug.verbose=datapath
./cilium-cli status --wait
./cilium-cli connectivity test --include-unsafe-tests  --flush-ct --test "skipall"  -v  -p

```

### 2. Check pod to remote nodeport connectivity

```
# kubectl -ncilium-test get po -owide | grep client2
client2-ccd7b8bdf-dtdng              1/1     Running   0          32m   10.244.3.159   kind-worker2         <none>           <none>

# kubectl -nkube-system get po -owide | grep cilium | grep kind-worker
cilium-2lwb2                                 1/1     Running   0          34m   172.19.0.5     kind-worker2         <none>           <none>
cilium-9bn77                                 1/1     Running   0          34m   172.19.0.4     kind-worker          <none>           <none>
cilium-envoy-6s69m                           1/1     Running   0          34m   172.19.0.5     kind-worker2         <none>           <none>
cilium-envoy-wjhfh                           1/1     Running   0          34m   172.19.0.4     kind-worker          <none>           <none>
cilium-operator-7b87f5b697-w88vj             1/1     Running   0          34m   172.19.0.4     kind-worker   

# client2 pod is on kind-worker2, so we choose a remote nodeport on kind-worker
# kubectl -nkube-system exec cilium-9bn77 -- cilium service list  | grep -i hostport
17   172.19.0.4:4000           HostPort       1 => 10.244.2.175:8080 (active)            
18   [fc00:c111::4]:4000       HostPort       1 => [fd00:10:244:2::fc99]:8080 (active)   
19   0.0.0.0:4000              HostPort       1 => 10.244.2.175:8080 (active)   
20   [::]:4000                 HostPort       1 => [fd00:10:244:2::fc99]:8080 (active) 

# kubectl -ncilium-test exec client2-ccd7b8bdf-dtdng  -- curl 172.19.0.4:4000 -I
HTTP/1.1 200 OK
X-Powered-By: Express
Vary: Origin, Accept-Encoding
Access-Control-Allow-Credentials: true
Accept-Ranges: bytes
Cache-Control: public, max-age=0
Last-Modified: Tue, 09 Jan 2024 12:57:12 GMT
ETag: W/"809-18cee4c6040"
Content-Type: text/html; charset=UTF-8
Content-Length: 2057
Date: Wed, 05 Jun 2024 03:38:14 GMT
Connection: keep-alive
Keep-Alive: timeout=5
```

### 3. Apply an L7 ingress policy and check connectivity again

```
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: allow-8080-ingress
spec:
  endpointSelector:
    matchLabels:
      kind: echo
  ingress:
  - fromEntities:
    - all
    toPorts:
    - ports:
      - port: "8080"
        protocol: TCP
      rules:
        http:
        - method: "GET"
          path: "/$"
```

```
kubectl -ncilium-test apply -f allow-8080-ingress.yaml
```

Then pod to remote nodeport can't be connected:

```
$ kubectl -ncilium-test exec client2-ccd7b8bdf-dtdng  -- curl 172.19.0.4:4000 -I -v
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 172.19.0.4:4000...
  0     0    0     0    0     0      0      0 --:--:--  0:00:09 --:--:--     0^C
```

### Cilium Version

```
Client: 1.16.0-dev 8cd748de7f 2024-05-20T11:00:02+09:00 go version go1.22.3 linux/amd64
Daemon: 1.16.0-dev 8cd748de7f 2024-05-20T11:00:02+09:00 go version go1.22.3 linux/amd64
```

### Kernel Version

Linux liangzc-l-PF4RDLEQ 6.5.0-1023-oem #24-Ubuntu SMP PREEMPT_DYNAMIC Tue May  7 14:26:31 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

Client Version: v1.30.0
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.4

### Regression

_No response_

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

The cause of issue is the missing revDNAT for proxy's reply.

The request path is: pod1(node1) -> eth0(node1) -> eth0 (node2) -> proxy(node2) -> pod2
As this is a to-hostport (nodeport) request, the packet will be DNAT-ed at from-netdev@eth0(node2).
Also please note that eth0(node2) -> proxy(node2) and proxy(node2) -> pod2 are different skbs because of proxy traffic hijack.

Let's look at what's happening to proxy(node2)'s reply, it's supposed to go through: proxy(node2) -> eth0(node2) -> eth0(node1) -> pod1(node1), to establish a TCP connection between proxy(node2) and pod1(node1). The expected revDNAT is awaiting at to-netdev@eth0(node2).

However, due to IPsec enabled, the proxy(node2)'s reply, which is a pod-to-pod traffic, will be marked 0xe00 at from_host@cilium_host, and be IPsec encrypted before reaching to-netdev@eth0(host2). When to-netdev@eth0(host2) see the reply packets, they are encrypted and impossible to do revDNAT. The proxy's reply will finally arrive at pod1(host1), with a wrong src IP, ending up with being dropped by kernel due to `SKB_DROP_REASON_NO_SOCKET`. 

I want to point out that at this stage, proxy(host2) hasn't sent TCP syn to pod2(host2), because that happens only if connection between pod1(host1) and proxy(host2) is established.

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by jschwinger233**
This PR adds a new revDNAT hook at to-host@cilium_net to fix a connectivity issue.

Fixes: #32897 

```release-note
Fixes broken pod-to-remote-hostport connectivity when IPsec is used with L7 ingress policy and KPR.
```

### CODE REVIEW

File: bpf/lib/nodeport.h [hunk 1]
```diff
@@ -3346,14 +3357,20 @@ lb_handle_health(struct __ctx_buff *ctx __maybe_unused, __be16 proto)
...
 
+/* handle_nat_fwd() handles revDNAT, fib_lookup_redirect, and bpf_snat for
+ * nodeport. If revdnat_only is set to 1, fib_lookup and bpf_snat are skipped.
+ * The typical use case of handle_nat_fwd(revdnat_only=1) is for handling reply
+ * traffic that requires revDNAT prior to wireguard/IPsec encryption.
+ */
 static __always_inline int
 handle_nat_fwd(struct __ctx_buff *ctx, __u32 cluster_id, __be16 proto,
-	       struct trace_ctx *trace __maybe_unused,
+	       __u32 revdnat_only, struct trace_ctx *trace __maybe_unused,
```
Comment (brb) [inline bpf/lib/nodeport.h]: Nit: maybe `bool` instead of `__u32`?

File: bpf/bpf_host.c [hunk 2]
```diff
@@ -1651,7 +1651,23 @@ int cil_to_host(struct __ctx_buff *ctx)
...
 	 */
 	ctx_change_type(ctx, PACKET_HOST);
-#endif
+
+# ifdef ENABLE_NODEPORT
+	if (!validate_ethertype(ctx, &proto)) {
+		ret = DROP_UNSUPPORTED_L2;
+		goto out;
+	}
+	/* handle_nat_fwd() tail calls in the majority of cases, so control
```
Comment (julianwiedmann) [inline bpf/bpf_host.c:1665]: I'm pretty sure that the inline/tail-call decision is currently based on expected program size in `to-netdev`. Maybe we can optimize this a bit further in the future, and then inline the logic here.

File: bpf/bpf_host.c [hunk 3]
```diff
@@ -1651,7 +1651,23 @@ int cil_to_host(struct __ctx_buff *ctx)
 	 * to mark as PACKET_OTHERHOST and drop.
 	 */
 	ctx_change_type(ctx, PACKET_HOST);
-#endif
+
+# ifdef ENABLE_NODEPORT
+	if (!validate_ethertype(ctx, &proto)) {
+		ret = DROP_UNSUPPORTED_L2;
+		goto out;
```
Comment (julianwiedmann) [inline bpf/bpf_host.c]: Trying to understand the scope / what traffic this new RevDNAT section captures. Can we limit the impact of it at all - and *only* handle the specific traffic type that we know is broken at the moment? Ideally everything else would still float into `to-netdev`, and get RevDNAT there as before.

File: bpf/bpf_host.c [hunk 4]
```diff
@@ -1651,7 +1651,23 @@ int cil_to_host(struct __ctx_buff *ctx)
 	 * to mark as PACKET_OTHERHOST and drop.
 	 */
 	ctx_change_type(ctx, PACKET_HOST);
-#endif
+
+# ifdef ENABLE_NODEPORT
+	if (!validate_ethertype(ctx, &proto)) {
+		ret = DROP_UNSUPPORTED_L2;
```
Comment (julianwiedmann) [inline bpf/bpf_host.c]: Do we need to actually drop this traffic? Or can we just skip the RevDNAT for it?

File: bpf/lib/nodeport.h [hunk 5]
```diff
@@ -3346,14 +3357,20 @@ lb_handle_health(struct __ctx_buff *ctx __maybe_unused, __be16 proto)
...
 	       __s8 *ext_err __maybe_unused)
 {
 	int ret = CTX_ACT_OK;
 
 	ctx_store_meta(ctx, CB_CLUSTER_ID_EGRESS, cluster_id);
+	ctx_store_meta(ctx, CB_REVDNAT_ONLY, revdnat_only);
```
Comment (julianwiedmann) [inline bpf/lib/nodeport.h]: To expand on Martynas' comment - my experience is that we shouldn't pass booleans through their own CB slots (the result looks like [this](https://github.com/cilium/cilium/blob/29029ddca1daba9354f5b42384777466ebf86a45/bpf/lib/l3.h#L107) - using three slots for three booleans).

So can we maybe turn this into a `CB_NAT_FLAGS` instead, and pass a `#define NAT_FLAGS_REVDNAT_ONLY (1 << 0)` ? Then this hopefully becomes much easier to extend later.

**Review by julianwiedmann — COMMENTED** [review]
Just dropping some initial thoughts, overall this looks great!

**Comment by julianwiedmann** [comment]
To test this, we'd need to extend https://github.com/cilium/cilium-cli/pull/2670 correct?

**Comment by jschwinger233** [comment]
@julianwiedmann yes, I just drafted the pr: https://github.com/cilium/cilium-cli/pull/2759

File: bpf/bpf_host.c (hunk 3 above)
Comment (jschwinger233) [reply bpf/bpf_host.c]: Make sense, I can wrap it with `if (skb->mark & 0xf00 == 0xe00)` so only to-be-IPsec-encrypted skbs are checking revDNAT.

File: bpf/bpf_host.c (hunk 2 above)
Comment (jschwinger233) [reply bpf/bpf_host.c:1665]: Interesting, yes.

File: bpf/lib/nodeport.h [hunk 6]
```diff
@@ -3216,9 +3227,11 @@ static __always_inline int
 handle_nat_fwd_ipv4(struct __ctx_buff *ctx, struct trace_ctx *trace,
 		    __s8 *ext_err)
 {
+	__u32 cb_nat_flags = ctx_load_meta(ctx, CB_NAT_FLAGS);
```
Comment (julianwiedmann) [inline bpf/lib/nodeport.h]: To be consistent with `handle_nat_fwd_ipv6()`:

```suggestion
	__u32 cb_nat_flags = ctx_load_and_clear_meta(ctx, CB_NAT_FLAGS);
```

**Review by julianwiedmann — COMMENTED** [review]
Looks good, one small note below.

I'm a bit out of the loop - now that the CLI development happens in `cilium/cilium`, should the connectivity test change be part of this PR? Maybe conditioned to `>= 1.17`, and then we relax it later once the exact fixed `v1.16.x` version is known?

**Comment by youngnick** [comment]
I don't think I can review this one properly, so I'm handing over to @jrajahalme.

**Review by julianwiedmann — APPROVED** [review]
looks good, thank you!

**Review by jrajahalme — APPROVED** [review]
LGTM

**Pull request merged**